### PR TITLE
Extract sample and meta data in Appsignal.Plug, whitelist request headers

### LINKED ISSUE #212: Whitelist request headers
Instead of sending all request headers, we'll add a whitelist (like [the one in the Ruby integration](https://github.com/appsignal/appsignal-ruby/blob/5757e28bac8043b9ac04c91b65130d0bb383ed33/lib/appsignal/transaction.rb#L17-L22)).

## Support

- https://app.intercom.io/a/apps/yzor8gyw/respond/inbox/540709/conversations/10287802391

### PR DESCRIPTION
**Pull request by jeffkreeftmeijer**
Closes #212.

### CODE REVIEW

File: lib/appsignal/plug.ex [hunk 1]
```diff
@@ -78,6 +78,41 @@ if Appsignal.plug? do
...
+          "peer" => peer(conn)
+        } |> Map.merge(extract_request_headers(conn))
+      }
+    end
+
+    def extract_request_headers(%Plug.Conn{req_headers: req_headers}) do
+      keys = ~w(
+        accept accept-charset accept-encoding accept-language cache-control
+        connection user-agent from referer range
+      )
```
Comment (jeffkreeftmeijer) [inline lib/appsignal/plug.ex]: I've based this patch on request headers that are expected to be set in a Phoenix request:

```
[{"host", "localhost:4001"},
 {"cookie", "__ar_v4=U35IKTLTJNEP7GWW6OH3N2%3A20161120%3A90%7CKJIDW6DV4NCMJHZA3MCKTP%3A20161120%3A90%7CQVGTR45XBFCQ3FOIW2JQJS%3A20161120%3A90; _session_id=3800dff4ac0321e743d74537d1da7d0d; autopilot_associated=true; visit_id=59688321b96e6f5d174dcc73; timeframe=7D; incident_status=open; accin-name="},
 {"connection", "keep-alive"}, {"upgrade-insecure-requests", "1"},
 {"accept", "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8"},
 {"user-agent",
  "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_3) AppleWebKit/602.4.8 (KHTML, like Gecko) Version/10.0.3 Safari/602.4.8"},
 {"referer", "http://localhost:4001/"}, {"accept-language", "en-us"},
 {"accept-encoding", "gzip, deflate"}]
```

There are optional headers (like the x-prefixed ones) that I haven't seen in Phoenix requests, so I'd say we don't add them until we need to. Is there anything you're missing specifically?

Comment (tombruijn) [inline lib/appsignal/plug.ex]: I think most x-headers are more common outside of development mode. Such as behind a loadbalancer (HTTP_X_FORWARDED_FOR), nginx (HTTP_X_QUEUE_TIME, HTTP_X_APPLICATION_START), Heroku (HTTP_X_HEROKU_QUEUE_WAIT_TIME), etc.

We can leave some out, but I think some other headers might still be useful for our users to get a better idea why a request errored out (missing header) or why it performed badly (long queue time).

Also, now that I look at it, it doesn't appear like we set the queue time for a request anywhere right now, do we? There's a`set_queue_start` function on Transaction, but I don't see that being called.

Comment (jeffkreeftmeijer) [inline lib/appsignal/plug.ex]: Added everything (except for `HTTP_X_MIDDLEWARE_START`) to the whitelist in 41c25c0. Could you open a new issue for `set_queue_start`?

File: lib/appsignal/plug.ex [hunk 2]
```diff
@@ -99,7 +99,14 @@ if Appsignal.plug? do
...
       accept accept-charset accept-encoding accept-language cache-control
-      connection user-agent from referer range
+      connection content-length user-agent from negotiate pragma referer range
+
+      auth-type gateway-interface path-translated remote-host remote-ident
+      remote-user remote-addr request-method server-name server-port
+      server-protocol request-uri path-info client-ip range
+
+      x-request-start x-queue-start x-queue-time x-heroku-queue-wait-time
+      x-application-start x-forwarded-for x-auth-token
```
Comment (tombruijn) [inline lib/appsignal/plug.ex]: The only one I'm not sure about is `x-auth-token` (same for the Ruby gem) that would undo the purpose of this PR, to filter out sensitive data such as API keys.

What's the reason that's in the Ruby gem, @matsimitsu @thijsc ? And can we leave it out here and in the Ruby gem? If our users want to save this data on a transaction I would say our users should set this manually with a tag/metadata.

**Pull request merged**